### PR TITLE
Accept worker queue list as a function

### PR DESCRIPTION
**Pull request by ThePletch**
This allows the Faktory worker to pull from a different set of queues on each job fetch, using whatever logic the user likes.

This change supports less-strict queue ordering, e.g. having two equal-priority queues that are ordered randomly on each fetch

### CODE REVIEW

**Comment by ThePletch** [comment]
(deleted one of the existing tests as part of this PR, since it appeared to test different functionality than its name, and was character-for-character identical with a test later in the file)

**Comment by ThePletch** [comment]
ooh we may need to migrate the "expand an empty queue list to `["default"]`" check to the getter function, since a function could return an empty array

**Comment by ThePletch** [comment]
@jbielick let me know if you need any additional docs or testing from this change

**Comment by ThePletch** [comment]
if you'd prefer this to be more deterministic and less open, the other option is to type queues as `(string | string[])[]`, where `['a', ['b', 'c']]` would always pull from `'a'` first, but would pull from `'b'` and `'c'` in either random order or round-robin order with each job fetch.

**Comment by jbielick** [comment]
@ThePletch Thanks for the PR. 

I suspect we should be `shuffle`ing the queues. I think that's probably supposed to be the default behavior and it's a regression because it used to be done: https://github.com/jbielick/faktory_worker_node/blob/cb0582de48575caf5c7a237cca80b47dc5659249/lib/processor.js#L91-L95

Should we change the default behavior? I think a custom queues function is an option, but it might be too much of an escape hatch at the moment.

**Comment by jbielick** [comment]
Now that I'm remembering more details, strict ordering _might_ be a good default to retain. The implementation of queue prioritization in Sidekiq is probably worth modeling here: https://phil.tech/2016/tips-on-sidekiq-queues/

What do you think about using the queue priority implementation from sidekiq here?

for example:
```js
queues: [
  ['critical', 10],
  ['high', 5],
  ['low', 1],
],
```

**Comment by ThePletch** [comment]
i think weighted priority is a workable feature to handle the use cases my org would like strict-but-equal ordering for, though i'd definitely expect a change to queue ordering to be launched as an opt-in feature unless you want this to be part of a new major version.

that said, i'm having trouble envisioning how this implementation would work within faktory's API, since faktory (as far as i can tell) expects a strictly ordered list of queues to fetch from as part of the request payload, and i'm hesitant to implement new features entirely client-side.

**Comment by ThePletch** [comment]
and in response to the original question: no, my team definitely still needs some mechanism for queue prioritization, so i think accepting a flat list of queues that gets shuffled with each fetch won't meet our needs.

**Comment by jbielick** [comment]
> i think weighted priority is a workable feature to handle the use cases my org would like strict-but-equal ordering for, though i'd definitely expect a change to queue ordering to be launched as an opt-in feature unless you want this to be part of a new major version.

I propose no change to the existing, default behavior. Queues are strictly ordered and would remain that way. No shuffling. Defining a list of queues with priorities would be the opt-in behavior.

> that said, i'm having trouble envisioning how this implementation would work within faktory's API, since faktory (as far as i can tell) expects a strictly ordered list of queues to fetch from as part of the request payload, and i'm hesitant to implement new features entirely client-side.

Faktory server would be unaware. We would simply be carefully ordering the queue names provided in the FETCH command. Here's some of the implementation in the sidekiq source: https://github.com/mperham/sidekiq/blob/7003aeb6336d3fdaed27f7444998a5891aef8f1b/lib/sidekiq.rb#L64-L81

So to summarize:

**Current Behavior:**

- The provided queues array is implicitly strictly ordered. Queue starvation may occur.
- No queue weighting is possible.

**Proposed Changes:**

- The provided queues array is implicitly strictly ordered. Queue starvation may occur. (no change here!)
- Queue weighting is possible by adding [queue, weight] tuples for the queues option.

Would that resolve the issues you have?

**Comment by ThePletch** [comment]
I think that approach will meet our needs. To make sure I understand the Sidekiq implementation we'll be mirroring here, it constructs a list of queues by repeating queues with a weight of `n` `n` times, thus increasing their likelihood to appear at the start of the queue when the queue list is shuffled proportionally?

**Comment by jbielick** [comment]
> I think that approach will meet our needs. To make sure I understand the Sidekiq implementation we'll be mirroring here, it constructs a list of queues by repeating queues with a weight of `n` `n` times, thus increasing their likelihood to appear at the start of the queue when the queue list is shuffled proportionally?

That's correct. Additionally, all queues still appear in the fetch, in random order as fallbacks:

```
FETCH winner other1 other2 othern...
```

^ where winner was "randomly selected" (it appears first), but they're all randomly ordered (due to the `suffle().unique()`) so if `winner` is an empty queue, queues are still processed in a weighted order because the server would send the first job from the queue list (in order).

Here are the docs I would add for it:

---

#### Queues

By default, faktory_worker_node will listen to a `default` queue. If you'd like to use multiple queues, specify them in `options.queues` when instantiating a `Worker` or via `--queue` when using the CLI.

Queues can be specified in two ways: **strictly ordered** and **weighted random**.

##### Strictly Ordered

To specify queues to be worked on in a strict order (empty one, then the next, etc), use an array of queue names for `options.queues`.

```js
faktory.work({ queues: ['immediately', 'afterwards'] })
```

In the previous example, all jobs are be fetched and started from the `immediately` queue before any are fetched and started from the `afterwards` queue.

##### Weighted Random

To specify queues to be worked on randomly, use an object of `[queueName: string]: number` entries. To ensure that some queues are processed more frequently than others, use a _higher_ weight number for that queue. Weights are relative. `100`, `200`, and `300` is no different than `1`, `2`, and `3`, respectively.

```js
faktory.work({ queues: { critical: 10, default: 1 } })
```

In the example above, the `critical` queue, weighted `10`, will have _roughly_ 10 jobs processed for every 1 job processed in the `default` queue. For equal weighting (just random queue processing), see below:

```js
faktory.work({ queues: { useast: 1, uswest: 1 } })
```

In the example above, `useast` and `uswest` have equal weights are worked on concurrently.
